### DISCUSSION TITLE
Adding custom actions

### DISCUSSION THREAD

**Discussion by cannycookie** [awcodes/filament-tiptap-editor#455] [Q&A]
The docs say we can override links and media actions but i need both of those.

I'm trying to add a custom action to add tokens to the editor.  Would it be possible to update the HasCustomActions concern to allow as many as we like with something like this:-

```php
<?php

namespace FilamentTiptapEditor\Concerns;

use Closure;
use Filament\Forms\Components\Actions\Action;

trait HasCustomActions
{
    /**
     * @var array<string, string|Closure|null>
     */
    protected array $customActions = [];

    /**
     * Set a custom action.
     *
     * @param  string  $name
     * @param  string|Closure  $action
     * @return static
     */
    public function setCustomAction(string $name, string | Closure $action): static
    {
        $this->customActions[$name] = $action;

        return $this;
    }

    /**
     * Get a custom action.
     *
     * @param  string  $name
     * @return Action
     */
    public function getCustomAction(string $name): Action
    {
        $action = $this->evaluate($this->customActions[$name] ?? null) ?? config("filament-tiptap-editor.{$name}_action");

        return $action::make();
    }

    /**
     * Alias methods for specific actions for backward compatibility.
     */
    public function linkAction(string | Closure $action): static
    {
        return $this->setCustomAction('link', $action);
    }

    public function mediaAction(string | Closure $action): static
    {
        return $this->setCustomAction('media', $action);
    }

    public function editMediaAction(string | Closure $action): static
    {
        return $this->setCustomAction('editMedia', $action);
    }

    public function getLinkAction(): Action
    {
        return $this->getCustomAction('link');
    }

    public function getMediaAction(): Action
    {
        return $this->getCustomAction('media');
    }

    public function getEditMediaAction(): Action
    {
        return $this->getCustomAction('editMedia');
    }
}
```

**Comment by awcodes** [comment]
What are you trying to accomplish? There's an inherit problem with trying to store any URLs with tokens in the database, because they will mostly likely be expired by the time they are read back out in the future. But that's an issue at a conceptual level and not an issue specifically with the plugin. Any wysiwyg type editor will have the same problem.

**Comment by cannycookie** [comment]
I want to add a custom button to the tool bar with a filament form but don't want to replace the link or media buttons.  Ultimately to allow the user to insert attribute placeholders like ``{{user.first_name}}`` within the editor.  My form has 1 select with a list of models and another select with a list of attributes.

**Comment by awcodes** [comment]
There's nothing stopping you from adding your own button with its own functionality.

**Comment by awcodes** [comment]
Sounds like you are just trying to make a merge tag though. But I'm sure I'm not fully understanding your use case.

**Comment by cannycookie** [comment]
Yes exactly a merge tag for templates

**Reply by awcodes** [reply]
So just use the merge tags. Not sure what that has to do with the media or link action.

**Reply by cannycookie** [reply]
I wanted to create a custom button to do this because we already have 67 merge tags in use and there are more required so i wanted to create a custom form that allows the user to choose the model and attribute with which to create the merge tag.

In the docs it says you can customise the link or media action but it's not clear how to add a custom action.  This is the point of what I added above so that a new action can be handled the HasCustomActions trait.  Does that make more sense?

**Reply by awcodes** [reply]
Ok. then you'll need to do a custom block, but there's no way to add it directly to the toolbar. Only custom extensions can be added directly to the toolbar.

Overriding the Link or Media action isn't going to be what you want in this case.

**Reply by cannycookie** [reply]
Is it viable to override the sidebar block which contains the merge tags?

**Reply by awcodes** [reply]
You're free to override any view you wish, it's just not recommended and you will have to be aware of any possible breaking changes with new releases.
